test/lib/mock-blpapi.ts

```diff
@@ -80,181 +82,202 @@ export class Session extends events.EventEmitter implements ISession {
-                this.sendServiceOpened(cid);
-            });
-            ipc.on(util.format('openService-%d-fail', cid), (): void => {
-                this.sendServiceOpenFailure(uri, cid);
-            });
+    subscribe(subs: Subscription[],
+              identity: IIdentity,
+              cb?: (err: any) => void): Promise<void>;
+
+    subscribe(subs: Subscription[],
+              label: string,
+              cb?: (err: any) => void): Promise<void>;
+
+    subscribe(subs: Subscription[],
+              identity: IIdentity,
+              label: string,
+              cb?: (err: any) => void): Promise<void>;
+
+    subscribe(subs: Subscription[],
+              arg1?: IIdentity | string | ((err: any) => void),
+              arg2?: string | ((err: any) => void),
+              arg3?: (err: any) => void): Promise<void>
+    {
+        var cb: (err: any) => void = undefined;
+        if (typeof arguments[arguments.length - 1] === 'function') {
+           cb = arguments[arguments.length - 1];
         }
-        return cid;
-    }
+        return new Promise<void>((resolve: Function, reject: Function): void => {
+            // For subscribe, only communicate via ipc is supported
+            ipc.emit('wait-to-subscribe', subs);
 
-    subscribe(subs: Subscription[], identity?: Identity, label?: string): Session {
-        // For subscribe, only communicate via ipc is supported
-        ipc.emit('wait-to-subscribe', subs);
-        _.forEach(subs, (sub: Subscription): void => {
-            var serviceUri = getServiceForSecurity(sub.security);
-            assert(SERVICE_TO_SUBSCRIPTION_EVENTS_MAP[serviceUri],
-                   'Invalid service name.');
-            var events: string[] = SERVICE_TO_SUBSCRIPTION_EVENTS_MAP[serviceUri];
-            _.forEach(events, (e: string): void => {
-                ipc.on(util.format('subscription-%d-%s', sub.correlation, e), (): void => {
-                    this.sendSubscriptionData(e, sub.correlation);
+            var hasInvalidServiceUri = subs.some((sub: Subscription): boolean => {
+                var serviceUri = getServiceForSecurity(sub.security);
+                if (!SERVICE_TO_SUBSCRIPTION_EVENTS_MAP[serviceUri]) {
+                    return true;
+                }
+                return false;
+            });
+            if (hasInvalidServiceUri) {
+                reject(new Error('Invalid service name'));
+                return;
+            }
+            subs.forEach((sub: Subscription): void => {
+                var cid = _.findKey(subs, (other: Subscription): boolean => {
+                    return sub === other;
+                });
+                var serviceUri = getServiceForSecurity(sub.security);
+                var events: string[] = SERVICE_TO_SUBSCRIPTION_EVENTS_MAP[serviceUri];
+                _.forEach(events, (e: string): void => {
+                    ipc.on(util.format('subscription-%s', e), (): void => {
```

Comment:
We need a way to differentiate different subscriptions through the ipc channel(`cid`). In the current setup, all subscriptions will emit data if one commands coming through ipc.